File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.py
def solution(players, callings):
    grade_hash = {p:idx for idx, p in enumerate(players)}
    
    for p in callings:
        # 추월한사람의 players에서의 인덱스 저장 - grade_hash 이용 -> O(1)
        idx = grade_hash[p]
        # players 스왑
        players[idx-1], players[idx] = players[idx], players[idx-1]
        # grade_hash 순위 갱신
        grade_hash[p] -= 1
        grade_hash[players[idx]] += 1
    
    return players
        

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.swift
import Foundation

func solution(_ players:[String], _ callings:[String]) -> [String] {
    var players = players
    var playerRankDict = [String: Int]()
    
    for (idx, player) in players.enumerated() {
        playerRankDict[player] = idx
    }
    
    for winnerName in callings {
        let winnerIdx = playerRankDict[winnerName]!
        let looserName = players[winnerIdx - 1]
        let looserIdx = playerRankDict[looserName]!
        
        playerRankDict[winnerName, default: 0] -= 1
        playerRankDict[looserName, default: 0] += 1
        
        players.swapAt(winnerIdx, looserIdx)
    }
    
    return players
}